distributed/tests/test_as_completed.py

```diff
@@ -168,3 +170,39 @@ def test_async_for_py2_equivalent(c, s, a, b):
 
     with pytest.raises(StopAsyncIteration):
         yield seq.__anext__()
+
+
+def test_as_completed_error(loop):
+    with cluster() as (s, [a, b]):
+        with Client(s['address'], loop=loop) as c:
+            x = c.submit(throws, 1)
+            y = c.submit(sleep, 1)
+
+            ac = as_completed([x, y])
+            result = list(ac)
+
+            assert result == [x, y]
+            assert x.status == 'error'
+            assert y.status == 'finished'
+
+
+def test_as_completed_with_results(loop):
+    with cluster() as (s, [a, b]):
+        with Client(s['address'], loop=loop) as c:
+            x = c.submit(throws, 1)
+            y = c.submit(sleep, 1)
+            z = c.submit(inc, 1)
+
+            ac = as_completed([x, y, z], with_results=True)
+            y.cancel()
+            res = list(ac)
+
+            futs, results = zip(*res)
+            assert futs == (y, x, z)
+            assert x.status == 'error'
+            assert y.status == 'cancelled'
+            assert z.status == 'finished'
+
+            assert isinstance(results[0], CancelledError)
+            assert isinstance(results[1][1], RuntimeError)
```

Comment:
Given that the default to `raise_errors` is True I'm surprised that we didn't get an exception above.  Why is this?

Reply:
The exception gets raised within the `track_future` coroutine, and is consequently trapped by `tornado`; I had played with raising in `__next__` / `__anext__` so that it would actually surface to the top, but couldn't figure out how to manage [keeping it within a coroutine](https://github.com/dask/distributed/issues/2065#issuecomment-402778890) at the same time.  

The only solution I had come up with was:
```python
    def _get(self):
        res = self.queue.get()
        if self.with_results:
            fut, result = res
            assert sync(self.loop, fut._result, raiseit=False) == result
            six.reraise(*result)
        return res

    def __next__(self):
        ...
        return self._get()
```

but I'm not sure how robust that is, or whether it even solves the race condition problem?

Reply:
Maybe we pass a boolean `erred` in the tuple that we put into `self.queue`.  Whenver we get from that queue and we see that erred is True then we raise if self.raise_errors.

```python
def __next__(self):
    ...
    future, result, erred = self.queue.get()
    if erred and self.raise_errors:
        raise result
```

Reply:
Should I worry about adjusting `__anext__`?

Reply:
Yes, you should probably do the same on both.  The async style tests, the ones that use `@gen_cluster` will cover the `__anext__` method.